taichi/llvm/llvm_runtime.h

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#include <cstddef>
+#include <memory>
+
+#include "taichi/llvm/llvm_device.h"
+#include "taichi/llvm/llvm_offline_cache.h"
+#include "taichi/llvm/snode_tree_buffer_manager.h"
+#include "taichi/llvm/llvm_context.h"
+#include "taichi/struct/snode_tree.h"
+#include "taichi/program/compile_config.h"
+
+#include "taichi/runtime/runtime.h"
+#include "taichi/system/threading.h"
+#include "taichi/system/memory_pool.h"
+
+#define TI_RUNTIME_HOST
+#include "taichi/program/context.h"
+#undef TI_RUNTIME_HOST
+
+namespace taichi {
+namespace lang {
+
+namespace cuda {
+class CudaDevice;
+}  // namespace cuda
+
+namespace cpu {
+class CpuDevice;
+}  // namespace cpu
+
+class LlvmRuntimeExecutor {
+ public:
+  LlvmRuntimeExecutor(CompileConfig &config, KernelProfilerBase *profiler);
+
+  /* ------------------------ */
+  /* ---- Initialization ---- */
+  /* ------------------------ */
+  void initialize_llvm_runtime_snodes(
```

Comment:
I'm still a bit concerned that we are exposing many public interfaces, a lot of which may not be relevant to AOT users. Is it possible to friend this class with `LlvmProgramImpl`?

Reply:
Yeah, that sounds way better than exposing a lot of JIT-specific public interfaces. 

Will rearrange these public/private interfaces, thanks!

Reply:
Let's start with: 
1. materialize_runtime()
2. initialize_llvm_runtime_snodes()
3. allocate_memory_ndarray()
4. check_runtime_error() // for AOT tests

Will expose more once necessary